File: app/src/main/java/org/gnucash/android/db/adapter/CommoditiesDbAdapter.java
package org.gnucash.android.db.adapter;

import static org.gnucash.android.db.DatabaseSchema.CommodityEntry;
import static org.gnucash.android.model.Commodity.USD;
import static org.gnucash.android.model.Commodity.getLocaleCurrencyCode;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteStatement;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.gnucash.android.R;
import org.gnucash.android.app.GnuCashApplication;
import org.gnucash.android.db.DatabaseHolder;
import org.gnucash.android.model.Commodity;

import java.util.Objects;

import timber.log.Timber;

/**
 * Database adapter for {@link Commodity}
 */
public class CommoditiesDbAdapter extends DatabaseAdapter<Commodity> {
    private Commodity defaultCommodity;

    /**
     * Opens the database adapter with an existing database
     *
     * @param holder Database holder
     */
    public CommoditiesDbAdapter(@NonNull DatabaseHolder holder) {
        this(holder, true);
    }

    /**
     * Opens the database adapter with an existing database
     *
     * @param holder     Database holder
     * @param initCommon initialize commonly used commodities?
     */
    public CommoditiesDbAdapter(@NonNull DatabaseHolder holder, boolean initCommon) {
        super(holder, CommodityEntry.TABLE_NAME, new String[]{
            CommodityEntry.COLUMN_FULLNAME,
            CommodityEntry.COLUMN_NAMESPACE,
            CommodityEntry.COLUMN_MNEMONIC,
            CommodityEntry.COLUMN_LOCAL_SYMBOL,
            CommodityEntry.COLUMN_CUSIP,
            CommodityEntry.COLUMN_SMALLEST_FRACTION,
            CommodityEntry.COLUMN_QUOTE_FLAG,
            CommodityEntry.COLUMN_QUOTE_SOURCE,
            CommodityEntry.COLUMN_QUOTE_TZ
        }, true);
        if (initCommon) {
            initCommon();
        } else {
            defaultCommodity = getDefaultCommodity();
        }
    }

    /**
     * initialize commonly used commodities
     */
    public void initCommon() {
        Commodity.AUD = Objects.requireNonNull(getCurrency("AUD"));
        Commodity.CAD = Objects.requireNonNull(getCurrency("CAD"));
        Commodity.CHF = Objects.requireNonNull(getCurrency("CHF"));
        Commodity.EUR = Objects.requireNonNull(getCurrency("EUR"));
        Commodity.GBP = Objects.requireNonNull(getCurrency("GBP"));
        Commodity.JPY = Objects.requireNonNull(getCurrency("JPY"));
        USD = Objects.requireNonNull(getCurrency("USD"));

        defaultCommodity = Commodity.DEFAULT_COMMODITY = getDefaultCommodity();
    }

    @Nullable
    public static CommoditiesDbAdapter getInstance() {
        return GnuCashApplication.getCommoditiesDbAdapter();
    }

    @Override
    protected @NonNull SQLiteStatement bind(@NonNull SQLiteStatement stmt, @NonNull final Commodity commodity) {
        bindBaseModel(stmt, commodity);
        stmt.bindString(1, commodity.getFullname());
        stmt.bindString(2, commodity.getNamespace());
        stmt.bindString(3, commodity.getMnemonic());
        if (commodity.getLocalSymbol() != null) {
            stmt.bindString(4, commodity.getLocalSymbol());
        }
        if (commodity.getCusip() != null) {
            stmt.bindString(5, commodity.getCusip());
        }
        stmt.bindLong(6, commodity.getSmallestFraction());
        stmt.bindLong(7, commodity.getQuoteFlag() ? 1 : 0);
        if (commodity.getQuoteSource() != null) {
            stmt.bindString(8, commodity.getQuoteSource());
        }
        if (commodity.getQuoteTimeZoneId() != null) {
            stmt.bindString(9, commodity.getQuoteTimeZoneId());
        }

        return stmt;
    }

    @Override
    public Commodity buildModelInstance(@NonNull final Cursor cursor) {
        String fullname = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_FULLNAME));
        String mnemonic = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_MNEMONIC));
        String namespace = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_NAMESPACE));
        String cusip = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_CUSIP));
        String localSymbol = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_LOCAL_SYMBOL));

        int fraction = cursor.getInt(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_SMALLEST_FRACTION));
        String quoteSource = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_QUOTE_SOURCE));
        String quoteTZ = cursor.getString(cursor.getColumnIndexOrThrow(CommodityEntry.COLUMN_QUOTE_TZ));

        Commodity commodity = new Commodity(fullname, mnemonic, namespace, fraction);
        populateBaseModelAttributes(cursor, commodity);
        commodity.setCusip(cusip);
        commodity.setQuoteSource(quoteSource);
        commodity.setQuoteTimeZone(quoteTZ);
        commodity.setLocalSymbol(localSymbol);

        return commodity;
    }

    @Override
    public Cursor fetchAllRecords() {
        return fetchAllRecords(CommodityEntry.COLUMN_MNEMONIC + " ASC");
    }

    /**
     * Fetches all commodities in the database sorted in the specified order
     *
     * @param orderBy SQL statement for orderBy without the ORDER_BY itself
     * @return Cursor holding all commodity records
     */
    public Cursor fetchAllRecords(String orderBy) {
        return mDb.query(mTableName, null, null, null, null, null, orderBy);
    }

    /**
     * Returns the commodity associated with the ISO4217 currency code
     *
     * @param currencyCode 3-letter currency code
     * @return Commodity associated with code or null if none is found
     */
    @Nullable
    public Commodity getCurrency(@Nullable String currencyCode) {
        if (TextUtils.isEmpty(currencyCode)) {
            return null;
        }
        if (isCached) {
            for (Commodity commodity : cache.values()) {
                if (commodity.isCurrency() && commodity.getCurrencyCode().equals(currencyCode)) {
                    return commodity;
                }
            }
        }
        String where = CommodityEntry.COLUMN_MNEMONIC + "=?"
            + " AND " + CommodityEntry.COLUMN_NAMESPACE + " IN ('" + Commodity.COMMODITY_CURRENCY + "','" + Commodity.COMMODITY_ISO4217 + "')";
        String[] whereArgs = new String[]{currencyCode};
        Cursor cursor = fetchAllRecords(where, whereArgs, null);
        try {
            if (cursor.moveToFirst()) {
                Commodity commodity = buildModelInstance(cursor);
                if (isCached) {
                    cache.put(commodity.getUID(), commodity);
                }
                return commodity;
            } else {
                String msg = "Commodity not found in the database: " + currencyCode;
                Timber.e(msg);
            }
        } finally {
            cursor.close();
        }

        switch (currencyCode) {
            case "AUD":
                return Commodity.AUD;
            case "CAD":
                return Commodity.CAD;
            case "CHF":
                return Commodity.CHF;
            case "EUR":
                return Commodity.EUR;
            case "GBP":
                return Commodity.GBP;
            case "JPY":
                return Commodity.JPY;
            case "USD":
                return USD;
            default:
                return null;
        }
    }

    public String getCommodityUID(String currencyCode) {
        Commodity commodity = getCurrency(currencyCode);
        return (commodity != null) ? commodity.getUID() : null;
    }

    public String getCurrencyCode(@NonNull String guid) {
        Commodity commodity = getRecord(guid);
        if (commodity != null) {
            return commodity.getCurrencyCode();
        }
        throw new IllegalArgumentException("Commodity not found");
    }

    @Nullable
    public Commodity loadCommodity(@NonNull Commodity commodity) {
        if (commodity.id != 0) {
            return commodity;
        }
        try {
            commodity = getRecord(commodity.getUID());
        } catch (Exception e) {
            // Commodity not found.
            commodity = getCurrency(commodity.getCurrencyCode());
        }
        return commodity;
    }

    @NonNull
    public Commodity getDefaultCommodity() {
        Commodity commodity = defaultCommodity;
        if (commodity != null) {
            return commodity;
        }

        Context context = holder.context;
        String prefKey = context.getString(R.string.key_default_currency);
        SharedPreferences preferences = getBookPreferences();
        String currencyCode = preferences.getString(prefKey, null);
        if (currencyCode == null) {
            currencyCode = getLocaleCurrencyCode();
        }
        defaultCommodity = commodity = getCurrency(currencyCode);
        return (commodity != null) ? commodity : Commodity.DEFAULT_COMMODITY;
    }

    public void setDefaultCurrencyCode(@Nullable String currencyCode) {
        Context context = holder.context;
        SharedPreferences preferences = getBookPreferences();
        String prefKey = context.getString(R.string.key_default_currency);
        preferences.edit().putString(prefKey, currencyCode).apply();

        Commodity commodity = getCurrency(currencyCode);
        if (commodity != null) {
            defaultCommodity = commodity;
            Commodity.DEFAULT_COMMODITY = commodity;
        }
    }
}

File: app/src/main/java/org/gnucash/android/db/adapter/CommoditiesDbAdapter.kt
package org.gnucash.android.db.adapter

import android.database.Cursor
import android.database.sqlite.SQLiteStatement
import androidx.core.content.edit
import org.gnucash.android.R
import org.gnucash.android.app.GnuCashApplication
import org.gnucash.android.db.DatabaseHolder
import org.gnucash.android.db.DatabaseSchema.CommodityEntry
import org.gnucash.android.db.bindBoolean
import org.gnucash.android.db.bindInt
import org.gnucash.android.db.getInt
import org.gnucash.android.db.getString
import org.gnucash.android.model.Commodity
import org.gnucash.android.model.Commodity.Companion.getLocaleCurrencyCode
import timber.log.Timber

/**
 * Database adapter for [Commodity]
 */
class CommoditiesDbAdapter(
    holder: DatabaseHolder,
    initCommon: Boolean = false
) : DatabaseAdapter<Commodity>(
    holder,
    CommodityEntry.TABLE_NAME,
    arrayOf(
        CommodityEntry.COLUMN_FULLNAME,
        CommodityEntry.COLUMN_NAMESPACE,
        CommodityEntry.COLUMN_MNEMONIC,
        CommodityEntry.COLUMN_LOCAL_SYMBOL,
        CommodityEntry.COLUMN_CUSIP,
        CommodityEntry.COLUMN_SMALLEST_FRACTION,
        CommodityEntry.COLUMN_QUOTE_FLAG,
        CommodityEntry.COLUMN_QUOTE_SOURCE,
        CommodityEntry.COLUMN_QUOTE_TZ
    ),
    true
) {
    private var _defaultCommodity: Commodity? = null

    /**
     * Opens the database adapter with an existing database
     *
     * @param holder     Database holder
     * @param initCommon initialize commonly used commodities?
     */
    /**
     * Opens the database adapter with an existing database
     *
     * @param holder Database holder
     */
    init {
        if (initCommon) {
            initCommon()
        }
    }

    /**
     * initialize commonly used commodities
     */
    fun initCommon() {
        Commodity.AUD = getCurrency("AUD")!!
        Commodity.CAD = getCurrency("CAD")!!
        Commodity.CHF = getCurrency("CHF")!!
        Commodity.EUR = getCurrency("EUR")!!
        Commodity.GBP = getCurrency("GBP")!!
        Commodity.JPY = getCurrency("JPY")!!
        Commodity.USD = getCurrency("USD")!!
        Commodity.DEFAULT_COMMODITY = defaultCommodity
    }

    override fun bind(stmt: SQLiteStatement, commodity: Commodity): SQLiteStatement {
        bindBaseModel(stmt, commodity)
        stmt.bindString(1, commodity.fullname)
        stmt.bindString(2, commodity.namespace)
        stmt.bindString(3, commodity.mnemonic)
        if (commodity.localSymbol != null) {
            stmt.bindString(4, commodity.localSymbol)
        }
        if (commodity.cusip != null) {
            stmt.bindString(5, commodity.cusip)
        }
        stmt.bindInt(6, commodity.smallestFraction)
        stmt.bindBoolean(7, commodity.quoteFlag)
        if (commodity.quoteSource != null) {
            stmt.bindString(8, commodity.quoteSource)
        }
        if (commodity.getQuoteTimeZoneId() != null) {
            stmt.bindString(9, commodity.getQuoteTimeZoneId())
        }

        return stmt
    }

    override fun buildModelInstance(cursor: Cursor): Commodity {
        val fullname = cursor.getString(CommodityEntry.COLUMN_FULLNAME)
        val mnemonic = cursor.getString(CommodityEntry.COLUMN_MNEMONIC)!!
        val namespace = cursor.getString(CommodityEntry.COLUMN_NAMESPACE)!!
        val cusip = cursor.getString(CommodityEntry.COLUMN_CUSIP)
        val localSymbol = cursor.getString(CommodityEntry.COLUMN_LOCAL_SYMBOL)

        val fraction = cursor.getInt(CommodityEntry.COLUMN_SMALLEST_FRACTION)
        val quoteSource = cursor.getString(CommodityEntry.COLUMN_QUOTE_SOURCE)
        val quoteTZ = cursor.getString(CommodityEntry.COLUMN_QUOTE_TZ)

        val commodity = Commodity(fullname, mnemonic, namespace, fraction)
        populateBaseModelAttributes(cursor, commodity)
        commodity.cusip = cusip
        commodity.quoteSource = quoteSource
        commodity.setQuoteTimeZone(quoteTZ)
        commodity.localSymbol = localSymbol

        return commodity
    }

    override fun fetchAllRecords(): Cursor {
        return fetchAllRecords(CommodityEntry.COLUMN_MNEMONIC + " ASC")
    }

    /**
     * Fetches all commodities in the database sorted in the specified order
     *
     * @param orderBy SQL statement for orderBy without the ORDER_BY itself
     * @return Cursor holding all commodity records
     */
    fun fetchAllRecords(orderBy: String?): Cursor {
        return db.query(tableName, null, null, null, null, null, orderBy)
    }

    /**
     * Returns the commodity associated with the ISO4217 currency code
     *
     * @param currencyCode 3-letter currency code
     * @return Commodity associated with code or null if none is found
     */
    fun getCurrency(currencyCode: String?): Commodity? {
        if (currencyCode.isNullOrEmpty()) {
            return null
        }
        if (isCached) {
            for (commodity in cache.values) {
                if (commodity.isCurrency && commodity.currencyCode == currencyCode) {
                    return commodity
                }
            }
        }
        val where = (CommodityEntry.COLUMN_MNEMONIC + "=?"
                + " AND " + CommodityEntry.COLUMN_NAMESPACE
                + " IN ('" + Commodity.COMMODITY_CURRENCY + "','" + Commodity.COMMODITY_ISO4217 + "')")
        val whereArgs = arrayOf<String?>(currencyCode)
        val cursor = fetchAllRecords(where, whereArgs, null) ?: return null
        try {
            if (cursor.moveToFirst()) {
                val commodity = buildModelInstance(cursor)
                if (isCached) {
                    cache[commodity.uid] = commodity
                }
                return commodity
            } else {
                val msg = "Commodity not found in the database: $currencyCode"
                Timber.e(msg)
            }
        } finally {
            cursor.close()
        }

        return when (currencyCode) {
            "AUD" -> Commodity.AUD
            "CAD" -> Commodity.CAD
            "CHF" -> Commodity.CHF
            "EUR" -> Commodity.EUR
            "GBP" -> Commodity.GBP
            "JPY" -> Commodity.JPY
            "USD" -> Commodity.USD
            else -> null
        }
    }

    fun getCommodityUID(currencyCode: String): String? {
        val commodity = getCurrency(currencyCode)
        return commodity?.uid
    }

    @Throws(IllegalArgumentException::class)
    fun getCurrencyCode(guid: String): String {
        val commodity =
            getRecordOrNull(guid) ?: throw IllegalArgumentException("Commodity not found")
        return commodity.currencyCode
    }

    fun loadCommodity(commodity: Commodity): Commodity {
        var commodity = commodity
        if (commodity.id != 0L) {
            return commodity
        }
        try {
            commodity = getRecord(commodity.uid)
        } catch (_: Exception) {
            // Commodity not found.
            commodity = getCurrency(commodity.currencyCode)!!
        }
        return commodity
    }

    val defaultCommodity: Commodity
        get() {
            var commodity: Commodity? = _defaultCommodity
            if (commodity != null) {
                return commodity
            }

            val context = holder.context
            val prefKey = context.getString(R.string.key_default_currency)
            val preferences = bookPreferences
            var currencyCode = preferences.getString(prefKey, null)
            if (currencyCode == null) {
                currencyCode = getLocaleCurrencyCode()
            }
            commodity = getCurrency(currencyCode) ?: Commodity.DEFAULT_COMMODITY
            _defaultCommodity = commodity
            return commodity
        }

    fun setDefaultCurrencyCode(currencyCode: String?) {
        val context = holder.context
        val preferences = bookPreferences
        val prefKey = context.getString(R.string.key_default_currency)
        preferences.edit { putString(prefKey, currencyCode) }

        val commodity = getCurrency(currencyCode)
        if (commodity != null) {
            _defaultCommodity = commodity
            Commodity.DEFAULT_COMMODITY = commodity
        }
    }

    companion object {
        val instance: CommoditiesDbAdapter get() = GnuCashApplication.commoditiesDbAdapter!!
    }
}
